README.md
2017年10月17日10:58:48
# 前链财经 Android  Dev Doc

## 简介
本文档旨在帮助开发者快速理解并基于 `前链财经 Android ` 端的开发，为二次开发提高效率。

整个项目相关的决策工作安排等记录在[thinksns-plus-document](https://github.com/zhiyicx/thinksns-plus-document).

整个项目代码风格都遵守[智艺创想移动端开发代码风格指南](https://github.com/zhiyicx/mobile-devices-code-style-guide)

## 工程基础配置说明

该工程使用 java 语言编写.支持 Android 4.0 (api 15) 以上系统.

IDE 为 Android Studio 2.2 编辑器.

Gradle 版本

```grovry
    distributionUrl=https\://services.gradle.org/distributions/gradle-2.14.1-all.zip

    dependencies {
        classpath 'com.android.tools.build:gradle:2.2.0'
    }
```
#### gradle 全局配置

**注：** 所有三方依赖包统一写入`config.gradle`，gradle配置文件`config.gradle`,需要在主项目的`buid.grale`中进行声明 `apply from："config.gralde"`

### Git 忽略文件说明

本工程忽略文件配置位于主工程下 `.gitignore`文件，可手动修改配置内容

### 代码混淆说明
开启代码混淆和混淆规则 `app` 的 `builde.gradle` 的文件下，`buildTypes` 节点添加 `release` 节点，` minifyEnabled` 属性表示是否开启混淆，`proguardFiles` 表示混淆依赖的文件，具体开启方法如下：
```
apply plugin: 'com.android.application'

android {
   ...
    buildTypes {
        release {
            minifyEnabled false
            proguardFiles getDefaultProguardFile('proguard-android.txt'), 'proguard-rules.pro'
        }
    }
    ...
}


```
混淆配置说明
```
 # 指定代码的压缩级别
-optimizationpasses 5
 # 是否使用大小写混合
-dontusemixedcaseclassnames
 # 是否混淆第三方jar
-dontskipnonpubliclibraryclasses
 # 混淆时是否做预校验
-dontpreverify
 # 混淆时是否记录日志
-verbose
 # 混淆时所采用的算法
-optimizations !code/simplification/arithmetic,!field/*,!class/merging/*
 # 保持哪些类不被混淆
-keep public class * extends android.content.ContentProvider
 # 保持哪些类不被混淆
-keep public class * extends android.app.backup.BackupAgentHelper
 # 保持哪些类不被混淆
-keep public class * extends android.preference.Preference
 # 保持哪些类不被混淆
-keep public class com.android.vending.licensing.ILicensingService
 # 保持 native 方法不被混淆
-keepclasseswithmembernames class * {
    native <methods>;
}
 # 保持自定义控件类不被混淆
-keepclasseswithmembers class * {
    public (android.content.Context, android.util.AttributeSet);
}
 # 保持自定义控件类不被混淆
-keepclasseswithmembers class * {
    public (android.content.Context, android.util.AttributeSet, int);
}
 # 保持自定义控件类不被混淆
-keepclassmembers class * extends android.app.Activity {
   public void *(android.view.View);
}
 # 保持枚举 enum 类不被混淆
-keepclassmembers enum * {
    public static **[] values();
    public static ** valueOf(java.lang.String);
}
 # 保持 Parcelable 不被混淆
-keep class * implements android.os.Parcelable {
  public static final android.os.Parcelable$Creator *;
}

```
> **注意:** 使用时请参考 `app/proguard-rules.pro`

## 项目安装、打包、签名说明
- [项目安装](document/tutorial/AndroidInstallDocTutorial.md)
- [创建签名文件说明](document/tutorial/AndroidCreateSignatureFileTutorial.md)
- [打包说明](document/tutorial/AndroidPackageTutorial.md)

## 文档位置说明

本工程文档统一记录在`ThinksnsPlus/document`文件夹下.

## 分支使用说明

分支命名方式以`git`工作流为准

示例:

```shell
master ( 主分支 )
develop (开发分支 )
feature/add_image  ( 特性分支 )
feature/add_image_jungle68 ( 个人开发可以加上作者 )
hotfix/pay_fail ( 修复分支 )

```

**注意:**

`作者分支`和`作者分支`的子分支按照作者意愿合并,由`作者分支`作者负责管理.

因为远端仓库存储了大量的主版本和子版本分支,为了减少不必要的更新和下载操作,所以`作者分支`不允许推送到远端.

为了方便后续使用自动化工具筛检`commit`内容,规定每次提交和`issues`相关的代码时,使用统一格式: 在提交信息末尾空一格,空一格后记录下`commit`类型,然后再在英文符号的括号内填写`issues code`.默认的 commit 类型为: 代码,文档以及测试

    示例代码:```(#9527) code 提交信息```
    示例代码:```(#9527) doc 提交信息```
    示例代码:```(#12 #30 #78) test 提交了测试信息```

## 框架选型

- 整体结构   [MVP+Dagger2](https://github.com/googlesamples/android-architecture/tree/todo-mvp-dagger/)

- 技术说明 ：  [retrofit](https://github.com/square/retrofit) + [dagger2](https://google.github.io/dagger/) + [rx](http://reactivex.io/) + [greenDao](https://github.com/greenrobot/greenDAO)


###   why choose this?


| 框架 | 劣势 | 优势 |
|:-------------:|:-------------:|:-------------|
|MVP|接口过多|1.官方推荐；<br><br>2.代码结构清晰;<br><br>3.代码解耦|
|retrofit||1.Square提供，质量高，不断维护更新；<br><br>2.易扩展，提供不同的Converter、Intercept等实现（也可以自定义），同时提供RxJava支持(返回Observable对象)，配合Jackson(或者Gson)和RxJava使用；<br><br>3.网络请求简洁明了
|dagger2||1.google官方推荐，配合mvp;<br><br>2.依赖注入，解耦代码；<br><br>3.简化对象的创建以及管理|
|rx||1.使用干净的输入/输出函数;<br><br>2.减少代码行数；<br><br>3.异步的错误处理；<br><br>4.便于线程操作|

### 工程结构

请查看[二次开发文档](document/tutorial/SecondaryDevelopmentTutorial.md)

CHANGELOG.md
# ThinkSNS+ 修改日志概要

## Version 1.8.0  *(2018-05-15)

新增

1.  视频;
2. 动态、相册 表增加 gif 标识， 图预览 持 gif 动画; 3. 最新好友、最新粉丝提示;
4. 503 服务端停机维护处 ;
5.  户拉 功能;
6. 三 账号解绑确认操作。
7. 服务 动态配置 表数据分 数 ;
8. 置顶审核 显示糖果与置顶时间;
9.  持超 4096  图查看
10. Android  持 gif 下载; 11. H5 基本聊天功能;


以下为本期已记录的部分内容


1. 应  图查看效果优化;
2. 对错误数据、空数据等脏数据兼容
3. 兼容服务 对图 的 mimetype 读取，防  件宽 为 0 的错误;
4. 修复创建圈 时的多 tag  法识别问题;
5. 修复创建圈 空数据时，设置封闭圈 崩溃问题;
6. 修复动态查看付费图  图查看后 表图 数据 法显示问题;
7. 隐藏没有好友的情况下的搜索;
8. 消息-通知 背景颜 优化;
9. 消息-聊天 背景颜 优化;
10. 选择好友 ，箭头与取消等设计统 ;
11. 清空历史消息后， 表  显示对话、 助 除外;
12. 聊天-智艺哥哥 可删除;没有数据的时候现在默认 字-“暂 消息”  显示时间;
13. 群信息中 查看群成员修改为查看成员;
14. 消息-聊天-群聊名字后 加上( 员数 );
15. 语 消息的播放图标优化;
16. 动态 字发送优化数据，去除 尾 效数据;
17. 去掉聊天发送 本，去除 尾 效数据;
18. 调整聊天点击+后弹出的菜单布局优化
19. 审核操作  案统 : 动态评论置顶-资讯评论置顶-帖 评论置顶-帖 置顶-圈 加 申请;
20. 应 中暂时 展示 视频通话功能;
21. 聊天消息中图 、位置消息背景圆 化;
22. 消息-聊天-聊天详情  图 发送后显示会有带过渡效果的闪烁，需要优化显示。
23. 动态发布 动画优化;
24. 群详情 顶部群名称增加群成员数 ;
25.  助 没有数据的时候默认排序到最后
26. 评论的、赞过的  的昵称和后 的 字增加 个空格,并调整资源内容 字  、颜 ;
27. 个 主 中，购买的 乐->移动到 乐中去(暂时去掉);
28. 优化环信的连接与重连;
29. 修复审核内容为空的时候，进 审核详情崩溃
30. 转让群主之后的转让群主消息错误;群主名字unicode码处 ;
31. 优化好友 表数据加载，防 本地的好友已经 是好友的情况导致  数据闪 下;
32. 修复聊天断开连接在连接 助 数据重复加载问题;
33. 调整 助 个 主 逻辑，去除  ，使 app个 主 ;
34. 优化动态，查看gif 图时，图 弹出时会闪 下;
35. 优化动态发布 查看 图点击缩回去的跳动效果;优化魅族 机发动态绑定账号等  的卡顿问题(硬件加速问题)
36. 魅族 机在发布动态选择图 ， 图 时会导致发布    渲染错误，  和背景 交错;并优化图选择后点击图 查看再点击确认的动画效果
37. 优化动态查看 图以及原图加载逻辑;
38. 优化动态查看 图、gif、 图、原图等
39. 优化录制视频的时候默摄像头;
40.  户被删除后，昵称显示改为-“该 户已被删除";
41. 消息-系统消息-外 消息的内容为系统消息的最后 条，但 保证外 的及时性是  的最新 条;
42. 消息-通知 所有 表的提示语都从服务 拉取最后 条，如果没有就显示缺省语;
43. 后端 1.8 版本发布切换到新的 auth 接 ，1.9 开始 tokens 接 废弃，前端需要替换;
44. 所有的 告图  有问题，整  告位推荐尺
45. 修复动态 表模拟 告 显示问题，并增加分享功能;
46. 消息-通知-系统通知 - 界 调整;
47. 去掉  视频的 窗 播放;
48. 修复查看审核通知偶发性异常;
49. 聊天定位发送的消息图  新;
50. 消息-聊天-选择好友 点击头像增加进 该 户的主 。
51. 选择封  顶部 Toolbar 间距调整;
52. 设置-意 反馈 检测是否已经输 内容时应该去除两边的空 字符，现在虽然空格 效，但能输 换行。
53. 我-粉丝/关注  表  的按钮可点击区域增 ，防 误触。
54.  告图倒计时显示优化，防 0s出现。
55. 消息-聊天-聊天详情  最后 条消息离输 的距离增 ，提示可视效果。
56.  助  新 (我-设置-意 反馈 当后台设置  助 时，点击和 助 聊天;未设置情况下，仍然使 系统的意 反馈功能。)
57. 修复看过圈 直接创建圈 没有圈 分类问题;
58. 修复新注册的 户，当 户没有糖果或者钱包时，发布 需要钱的内容时，错误跳转充值
59. 修复我-认证-个 认证 在选择图 的地 ，魅族 机返回时出现 块。
60. 修复单张 图本地发布后显示 度错误;
61. 兼容登录 登录按钮旁的加载动画偶发性的 显示;
62. 我-我的糖果 糖果商城下 的灰  块加 。
63. 我-我的糖果 充值糖果选项下 的那条线，修改为ios设计图样式。

richtexteditorlib/src/main/assets/markdown/richeditor.js
'use strict';
'use struct';

function HashMap(){
     /** Map 大小 **/
     var size = 0;
     /** 对象 **/
     var entry = new Object();

     /** 存 **/
     this.put = function (key , value)
     {
         if(!this.containsKey(key))
         {
             size ++ ;
         }
         entry[key] = value;
     }

     /** 取 **/
     this.get = function (key)
     {
         if( this.containsKey(key) )
         {
             return entry[key];
         }
         else
         {
             return null;
         }
     }

     /** 删除 **/
     this.remove = function ( key )
     {
         if( delete entry[key] )
         {
             size --;
         }
     }

     /** 是否包含 Key **/
     this.containsKey = function ( key )
     {
         return (key in entry);
     }

     /** 是否包含 Value **/
     this.containsValue = function ( value )
     {
         for(var prop in entry)
         {
             if(entry[prop] == value)
             {
                 return true;
             }
         }
         return false;
     }

     /** 所有 Value **/
     this.values = function ()
     {
         var values = new Array(size);
         for(var prop in entry)
         {
             values.push(entry[prop]);
         }
         return values;
     }

     /** 所有 Key **/
     this.keys = function ()
     {
         var keys = new Array(size);
         for(var prop in entry)
         {
             keys.push(prop);
         }
         return keys;
     }

     /** Map Size **/
     this.size = function ()
     {
         return size;
     }
 }

var RE = {
	currentRange: {
		startContainer: null,
		startOffset: 0,
		endContainer: null,
		endOffset: 0
	},
	cache: {
		editor: null,
		title: null,
		currentLink: null,
		line: null
	},
	commandSet: ['bold', 'italic', 'strikethrough', 'redo', 'undo'],
	schemeCache: {
		FOCUS_SCHEME: 'focus://',
		CHANGE_SCHEME: 'change://',
		STATE_SCHEME: 'state://',
		CALLBACK_SCHEME: 'callback://',
		IMAGE_SCHEME: 'image://'
	},
	setting: {
		screenWidth: 0,
		margin: 20
	},
	titleLimit:{
        txtNote:null,//文本框
        txtLimit:null,//提示字数的input
        limitCount:20,//限制的字数
        txtlength:0//到达限制时，字符串的长度
	},

	imageCache: new HashMap(),
	init: function init() {
		//初始化内部变量
		var _self = this;
		_self.initCache();
		_self.initSetting();
		_self.initLimit();
		_self.bind();
		_self.focus();
        console.log("init:::" + document.documentElement.outerHTML);
	},

	bind: function bind() {
		var _self = this;

		var _self$schemeCache = _self.schemeCache,
		    FOCUS_SCHEME = _self$schemeCache.FOCUS_SCHEME,
		    STATE_SCHEME = _self$schemeCache.STATE_SCHEME,
		    CALLBACK_SCHEME = _self$schemeCache.CALLBACK_SCHEME;
		document.addEventListener('selectionchange', _self.saveRange, false);

		_self.cache.title.addEventListener('focus', function () {
			AndroidInterface.setViewEnabled(true);
		}, false);

		_self.cache.title.addEventListener('blur', function () {
			AndroidInterface.setViewEnabled(false);
		}, false);

		_self.cache.editor.addEventListener('blur', function () {
			_self.saveRange();
		}, false);

		_self.cache.editor.addEventListener('click', function (evt) {
			_self.saveRange();
			_self.getEditItem(evt);
		}, false);

		_self.cache.editor.addEventListener('keyup', function (evt) {
			if (evt.which == 37 || evt.which == 39 || evt.which == 13 || evt.which == 8) {
				_self.getEditItem(evt);
			}
		}, false);

		_self.cache.editor.addEventListener('input', function () {
			AndroidInterface.setHtmlContent(_self.cache.title.innerHTML.length , _self.markdownWords().length);
		}, false);

		_self.titleLimit.txtNote.addEventListener('input', function () {
			_self.wordsLimit();
		}, false);

		_self.cache.title.addEventListener('input', function () {
			var content=_self.markdownWords();
			AndroidInterface.setHtmlContent(_self.cache.title.innerHTML.length , content.length);
		}, false);
	},

	initCache: function initCache() {
		var _self = this;
		_self.cache.editor = document.getElementById('editor');
		_self.cache.title = document.getElementById('title');
		_self.cache.line = document.getElementsByClassName('line')[0];
		_self.cache.editor.style.minHeight = window.innerHeight - 69 + 'px';
	},
	initSetting: function initSetting() {
		var _self = this;
		_self.setting.screenWidth = window.innerWidth - 20;
	},
	initLimit: function initLimit(){
	    var _self = this;
	    _self.titleLimit.txtNote = document.getElementById("title");
	    _self.titleLimit.txtLimit = document.getElementById("txtCount");
	    // 屏蔽了标题的回车输入
	    _self.titleLimit.txtNote.onkeypress = function(){
                                        if (event.keyCode == 13 ){
                                            return false;
                                        }

                                    }
	},
	focus: function focus() {
		//聚焦
		var _self = this;
		var range = document.createRange();
		range.selectNodeContents(this.cache.title);
		range.collapse(false);
		var select = window.getSelection();
		select.removeAllRanges();
		select.addRange(range);
		_self.cache.title.focus();
	},
	getHtml: function getHtml() {
		var _self = this;
		return _self.cache.editor.innerHTML;
	},
	noMarkdownWords: function noMarkdownWords() {
		var _self = this;
		var content = _self.cache.editor.innerHTML.replace(/<(?!\/?a)\/?[^>]*>|\u56FE\u7247\u4E0A\u4F20\u5931\u8D25\uFF0C\u8BF7\u70B9\u51FB\u91CD\u8BD5|\s+/g, '');
        AndroidInterface.noMarkdownWords(content);
		return content;
	},
	markdownWords: function markdownWords() {
        var _self = this;
//        var content = _self.cache.editor.innerHTML.replace(/<div\\s+\\S+>\\s+\\S+<\/div>|<[divimginput]+ class=".*">|\u56FE\u7247\u4E0A\u4F20\u5931\u8D25\uFF0C\u8BF7\u70B9\u51FB\u91CD\u8BD5/g, '').replace(/\n|\t/g,'').replace(/<div>[u4e00-u9fa5]+<\/div>/g,"").replace(/div|span/g,'p').trim();
        // 内容返回出去后 由 Java 处理
        var content = _self.cache.editor.innerHTML;
        AndroidInterface.markdownWords(content);
        return content;
    },

    changeTitlePlaceholder: function changeTitlePlaceholder(value){
        this.cache.title.setAttribute("title-placeholder", value);
    },

    changeContentPlaceholder: function changeContentPlaceholder(value){
        this.cache.editor.setAttribute("editor-placeholder", value);
    },

    getTitle: function getTitle() {
        var _self = this;
        var title = _self.cache.title.innerHTML.replace(/<br>/g,'');
        return title;
    },

    resultWords: function resultWords(boolean) {
        var _self = this;
        var value = boolean;
        console.log("isPublish:::" + value);
        AndroidInterface.resultWords(_self.getTitle(),_self.markdownWords(),_self.noMarkdownWords(),document.documentElement.outerHTML,value);
    },

    // 暂时没用到
    restoreDraft: function restoreDraft(title,content){
        var _self = this;
        _self.cache.title.innerHTML = title;
    },

	saveRange: function saveRange() {
		//保存节点位置
		var _self = this;
		var selection = window.getSelection();
		if (selection.rangeCount > 0) {
			var range = selection.getRangeAt(0);
			var startContainer = range.startContainer,
			    startOffset = range.startOffset,
			    endContainer = range.endContainer,
			    endOffset = range.endOffset;

			_self.currentRange = {
				startContainer: startContainer,
				startOffset: startOffset,
				endContainer: endContainer,
				endOffset: endOffset
			};
		}
	},
	reduceRange: function reduceRange() {
		//还原节点位置
		var _self = this;
		var _self$currentRange = _self.currentRange,
		    startContainer = _self$currentRange.startContainer,
		    startOffset = _self$currentRange.startOffset,
		    endContainer = _self$currentRange.endContainer,
		    endOffset = _self$currentRange.endOffset;

		var range = document.createRange();
		var selection = window.getSelection();
		selection.removeAllRanges();
		range.setStart(startContainer, startOffset);
		range.setEnd(endContainer, endOffset);
		selection.addRange(range);
	},
	exec: function exec(command) {
		//执行指令
		var _self = this;
		if (_self.commandSet.indexOf(command) !== -1) {
		    console.log("exec:::" + command);
			document.execCommand(command, false, null);
			console.log("execCommand:::" + document.documentElement.outerHTML);
		} else {
			var value = '<' + command + '>';
			document.execCommand('formatBlock', false, value);
			_self.getEditItem({});
		}
	},

	// 同步状态
	getEditItem: function getEditItem(evt) {
		//通过点击时，去获得一个当前位置的所有状态
		var _self = this;
		var _self$schemeCache2 = _self.schemeCache,
		    STATE_SCHEME = _self$schemeCache2.STATE_SCHEME,
		    CHANGE_SCHEME = _self$schemeCache2.CHANGE_SCHEME,
		    IMAGE_SCHEME = _self$schemeCache2.IMAGE_SCHEME;
		if (evt.target && evt.target.tagName === 'A') {
			_self.cache.currentLink = evt.target;
			var name = evt.target.innerText;
			var href = evt.target.getAttribute('href');
			window.location.href = CHANGE_SCHEME + encodeURI(name + '@_@' + href);
		} else {
			if (evt.which == 8) {
				AndroidInterface.noMarkdownWords(_self.noMarkdownWords());
			}
			var items = [];
			_self.commandSet.forEach(function (item) {
				if (document.queryCommandState(item)) {
				    console.log("queryCommandState:::" + item);
					items.push(item);
				}
			});
			console.log("getEditItem:::" + document.documentElement.outerHTML);
			if (document.queryCommandValue('formatBlock')) {
				items.push(document.queryCommandValue('formatBlock'));
			}
			window.location.href = STATE_SCHEME + encodeURI(items.join(','));
		}
	},
	insertHtml: function insertHtml(html) {
		var _self = this;
		document.execCommand('insertHtml', false, html);
	},

	insertHtmlDIV: function insertHtmlDIV(html) {
	    console.log("insertHtmlDIV:::" + html);
        document.getElementById('editor').innerHTML=html;
    },

	setBackgroundColor: function setBackgroundColor(color) {
	    var _self = this;
	    document.body.style.backgroundColor = color;
	},
	setFontColor: function setFontColor(color) {
        document.body.style.color = color;
	},
	setLineColor: function setLineColor(color) {
	    var _self = this;
	    _self.cache.editor.style.borderColor = color;
	},

	// 插入分割线
	insertLine: function insertLine() {
		var _self = this;
		var html = '<hr><div><br></div>';
		_self.insertHtml(html);
		_self.getEditItem({});
		window.scrollTo(0,document.body.scrollHeight);
	},

	// 插入链接
	insertLink: function insertLink(name, url) {
		var _self = this;
		var html = '<a href="' + url + '" class="editor-link">' + name + '</a>';
		_self.insertHtml(html);
	},

	// 修改链接
	changeLink: function changeLink(name, url) {
		var _self = this;
		var current = _self.cache.currentLink;
		var len = name.length;
		current.innerText = name;
		current.setAttribute('href', url);
		var selection = window.getSelection();
		var range = selection.getRangeAt(0).cloneRange();
		var _self$currentRange2 = _self.currentRange,
		    startContainer = _self$currentRange2.startContainer,
		    endContainer = _self$currentRange2.endContainer;

		selection.removeAllRanges();
		range.setStart(startContainer, len);
		range.setEnd(endContainer, len);
		selection.addRange(range);
	},

	pareseHtml: function pareseHtml(title,content){
	    var _self = this;
        _self.cache.title.innerHTML = title;
	    var image = '<div><br></div><div class="block">\n\t\t\t\t<div class="img-block"><div style="width: 100% " class="process">\n\t\t\t\t\t<div class="fill">\n\t\t\t\t\t</div>\n\t\t\t\t</div>\n\t\t\t\t<img class="images" data-id="' + id + '" style="width: 100% ; height: auto ;" src="' + url + '"/>\n\t\t\t\t<div class="cover" style="width: 100% ; height: auto "></div>\n\t\t\t\t<div class="delete">\n\t\t\t\t\t<img class="error" src="./reload.png">\n\t\t\t\t\t<div class="tips">\u56FE\u7247\u4E0A\u4F20\u5931\u8D25\uFF0C\u8BF7\u70B9\u51FB\u91CD\u8BD5</div>\n\t\t\t\t\t<div class="markdown"></div>\n\t\t\t\t</div></div>\n\t\t\t\t<input class="dec" type="text" placeholder="\u8BF7\u8F93\u5165\u56FE\u7247\u540D\u5B57">\n\t\t\t</div><div><br></div>';

	},

	// 插入图片
	insertImage: function insertImage(url, id, width, height) {
		var _self = this;
		var newWidth = 0,
		    newHeight = 0;
		var screenWidth = _self.setting.screenWidth;

		if (width > screenWidth) {
			newWidth = screenWidth;
			newHeight = height * newWidth / width;
		} else {
			newWidth = width;
			newHeight = height;
		}
		var image = '<div><br></div><div class="block">\n\t\t\t\t<div class="img-block"><div style="width: ' + newWidth + 'px" class="process">\n\t\t\t\t\t<div class="fill">\n\t\t\t\t\t</div>\n\t\t\t\t</div>\n\t\t\t\t<img class="images" data-id="' + id + '" style="width: ' + newWidth + 'px; height: ' + newHeight + 'px;" src="' + url + '"/>\n\t\t\t\t<div class="cover" style="width: ' + newWidth + 'px; height: ' + newHeight + 'px"></div>\n\t\t\t\t<div class="delete">\n\t\t\t\t\t<img class="error" src="./reload.png">\n\t\t\t\t\t<div class="tips">\u56FE\u7247\u4E0A\u4F20\u5931\u8D25\uFF0C\u8BF7\u70B9\u51FB\u91CD\u8BD5</div>\n\t\t\t\t\t<div class="markdown"></div>\n\t\t\t\t</div></div>\n\t\t\t\t<input class="dec" type="text" placeholder="\u8BF7\u8F93\u5165\u56FE\u7247\u540D\u5B57">\n\t\t\t</div><div><br></div>';
		if( width * height <= 0 ){
		    image = '<div><br></div><div class="block">\n\t\t\t\t<div class="img-block"><div style="width: 100% " class="process">\n\t\t\t\t\t<div class="fill">\n\t\t\t\t\t</div>\n\t\t\t\t</div>\n\t\t\t\t<img class="images" data-id="' + id + '" style="width: 100% ; height: auto ;" src="' + url + '"/>\n\t\t\t\t<div class="cover" style="width: 100% ; height: auto "></div>\n\t\t\t\t<div class="delete">\n\t\t\t\t\t<img class="error" src="./reload.png">\n\t\t\t\t\t<div class="tips">\u56FE\u7247\u4E0A\u4F20\u5931\u8D25\uFF0C\u8BF7\u70B9\u51FB\u91CD\u8BD5</div>\n\t\t\t\t\t<div class="markdown"></div>\n\t\t\t\t</div></div>\n\t\t\t\t<input class="dec" type="text" placeholder="\u8BF7\u8F93\u5165\u56FE\u7247\u540D\u5B57">\n\t\t\t</div><div><br></div>';
		}
		_self.insertHtml(image);
		var img = document.querySelector('img[data-id="' + id + '"]');
		var imgBlock = img.parentNode;
		imgBlock.parentNode.contentEditable = false;
		imgBlock.addEventListener('click', function (e) {
			e.stopPropagation();
			var current = e.currentTarget;
			var img = current.querySelector('.images');
			var id = img.getAttribute('data-id');
			window.location.href = _self.schemeCache.IMAGE_SCHEME + encodeURI(id);
		}, false);
        imgBlock.parentNode.addEventListener('DOMNodeRemovedFromDocument', function (evt) {
            evt.stopPropagation();
            var current = evt.currentTarget;
            var img = current.querySelector('.images');
            var id = img.getAttribute('data-id');
            AndroidInterface.deleteImage(id);
        }, false);
		_self.imageCache.put(id, imgBlock.parentNode);
		window.scrollTo(0,document.body.scrollHeight);
	},

	addImageClickListener: function addImageClickListener(ids){
        var _self = this;
        console.log("addImageClickListener:::" + ids);
        var array = ids.split(',');

        for(var item in array) {
            var img = document.querySelector('img[data-id="' + array[item] + '"]');
            if(img != null){
                console.log("addImageClick:::" + array[item]);
                var imgBlock = img.parentNode;
                imgBlock.parentNode.contentEditable = false;
                imgBlock.addEventListener('click', function (e) {
                    e.stopPropagation();
                    var current = e.currentTarget;
                    var img = current.querySelector('.images');
                    var id = img.getAttribute('data-id');
                    window.location.href = _self.schemeCache.IMAGE_SCHEME + encodeURI(id);
                }, false);
                imgBlock.parentNode.addEventListener('DOMNodeRemovedFromDocument', function (evt) {
                    evt.stopPropagation();
                    var current = evt.currentTarget;
                    var img = current.querySelector('.images');
                    var id = img.getAttribute('data-id');
                    AndroidInterface.deleteImage(id);
                }, false);
                _self.imageCache.put(array[item], imgBlock.parentNode);
            }
        }
	},

	// 图片上传进度
	changeProcess: function changeProcess(id, process,imageId) {
		var _self = this;
		var block = _self.imageCache.get(id);
		var fill = block.querySelector('.fill');
		fill.style.width = process + '%';
		if (process == 100) {
			var cover = block.querySelector('.cover');
			var dec = block.querySelector('.dec');
			var markdown = block.querySelector('.markdown');
			var process = block.querySelector('.process');
			var imgBlock = block.querySelector('.img-block');
			dec.addEventListener('input', function () {
                markdown.innerHTML="@!["+dec.value+"]("+imageId+")"
            }, false);

            if(markdown.innerHTML==""){
                markdown.innerHTML="@![image]("+imageId+")"
            }
			imgBlock.removeChild(cover);
			imgBlock.removeChild(process);
		}
	},

	hideTitle: function hideTitle(){
	    var _self = this;
	    console.log("hideTitle:::");
	    _self.cache.title.style.display="none";
	    _self.cache.line.style.display="none";
	    document.getElementById("content").style.paddingTop="0px";
	},

	// 限制标题输入字数
    wordsLimit: function wordsLimit(){
        var _self = this;
        var noteView = _self.titleLimit.txtNote;
        var limitCount = _self.titleLimit.limitCount;
        var InPutView = document.getElementById("title");
        if(InPutView.innerHTML.length < 10 ){
            document.getElementById("stay").style.display="none";
            return
        }
        if(InPutView.innerHTML.length >=10 ){
            document.getElementById("stay").style.display="block";
            document.getElementById("stay").style.color="green";
        }
        if(InPutView.innerHTML.length > 15 ){
            document.getElementById("stay").style.color="red";
        }
        if(noteView.innerHTML.length > limitCount ){
            noteView.innerHTML = noteView.innerHTML.substring(0,limitCount);
        }
        _self.titleLimit.txtLimit.innerText=noteView.innerHTML.length;
        _self.titleLimit.txtlength = noteView.innerHTML.length;//记录每次输入后的长度
    },

    // 删除图片
	removeImage: function removeImage(id) {
		var _self = this;
		var block = _self.imageCache.get(id);
		block.parentNode.removeChild(block);
		_self.imageCache.remove(id);
	},

	// 图片上传失败
	uploadFailure: function uploadFailure(id) {
		var _self = this;
		var block = _self.imageCache.get(id);
		var del = block.querySelector('.delete');
		del.style.display = 'block';
	},

	// 图片上传后重传
	uploadReload: function uploadReload(id) {
		var _self = this;
		var block = _self.imageCache.get(id);
		var del = block.querySelector('.delete');
		del.style.display = 'none';
	},

	stopBubble: function stopBubble(e) {
        //如果提供了事件对象，则这是一个非IE浏览器
        if ( e && e.stopPropagation ){
            e.stopPropagation(); //因此它支持W3C的stopPropagation()方法
        }else{
            window.event.cancelBubble = true; //否则，我们需要使用IE的方式来取消事件冒泡
        }
     },
};
RE.init();

